### PR TITLE
prototype of jekyll hooks, encapsulated

### PR DESCRIPTION
**Pull request by stevecrozz**
This is a second attempt at #3498 (adding hooks to Jekyll). In my last attempt @envygeeks had some complaints about the lack of encapsulation. This PR is an attempt to address those concerns and try out a new idea. This is definitely not ready to merge, but I wanted to get some eyes on it before I implement and write tests for the whole thing. The hooks concept is inspired by, but in this instance not perfectly compatible with octopress-hooks, so it's a bit of a departure from #3392 by @parkr, but I think it nicely encapsulates the hook logic.

Here's a practical example of how you could use one of these hooks to do something useful: https://github.com/stevecrozz/featherblade

### CODE REVIEW

**Comment by parkr** [comment]
Don't worry about the circleci failure.

This looks pretty good! @envygeeks?

**Comment by envygeeks** [comment]
:-1: I like it.

**Comment by envygeeks** [comment]
Sorry about the :-1: I actually meant to do :+1:

**Comment by stevecrozz** [comment]
Great. I'll flesh out the rest of it and add some test coverage.

**Comment by stevecrozz** [comment]
This is a bit more fleshed out now, although still lacking in tests and docs. I did do some one-off tests and it all seems to be working pretty well. Every hook receives the object being "hooked into" as the first argument.

One major difference from octopress-hooks worth pointing out is how I'm dealing with the payload objects just before rendering.

In octopress-hooks, there's a special hook for merging in new payload params. But I took a different approach and just pass the payload as an argument to each of the pre_render hooks. Anyone who implements a pre_render hook will receive the payload object and can modify it there if necessary as well as do any other things that need doing before rendering. pre_render is the only hook that receives a second argument at the moment. @imathis: let me know if you have any thoughts on this, especially if you already rejected this idea for some reason.

Next up is tests and docs, but I'll give this PR a few days to bake in case there's more feedback.

**Comment by stevecrozz** [comment]
I'm working on cucumber tests now. How thorough do these tests need to be? Do we want each hook for every container to have its own test scenario?

**Comment by envygeeks** [comment]
I say just write them and if there is a problem we'll point it out!

**Comment by stevecrozz** [comment]
@envygeeks / @parkr I've got cucumber tests and docs written. Do you have any feedback?

File: site/_docs/plugins.md [hunk 1]
```diff
@@ -469,6 +469,179 @@ module Jekyll
 end
 {% endhighlight %}
 
+## Hooks
+
```
Comment (parkr) [inline site/_docs/plugins.md]: Can you add an "unreleased" note here? Check out #3556 for an example.

**Comment by parkr** [comment]
Unless you have some thoughts, @envygeeks, let's merge and get this out there in a new beta so people can start building for it.

**Comment by stevecrozz** [comment]
Amended commit to include "unreleased" note in the docs

**Comment by alfredxing** [comment]
:+1: Nice work on this! We're missing hooks for Documents but that shouldn't be a huge issue for now.

**Comment by imathis** [comment]
I'm sorry I'm a bit late to this.

Unless I've overlooked it, this does not include:
- Site: pre_read, merge_payload
- Page/Post: merge_payload

For reference here's Octopress's [hook timeline](https://github.com/octopress/hooks#hook-timeline).

**Comment by stevecrozz** [comment]
@imathis That's correct. I removed pre_read from my PR while writing the test cases because I couldn't come up with a use case to test that wasn't already covered. There may be some good ones, but I didn't think of any.

As for merge_payload, I did depart from the octopress-hooks approach which is why I was interested in your feedback. I tried to cover the use case of altering the payload using pre_render which passes the payload to the callback as its second parameter. There's less code and less to test in this approach and it seems more flexible, but I'd be interested to know if I missed a good reason to implement it the way you already did.

**Comment by parkr** [comment]
@imathis ^^?

**Comment by imathis** [comment]
I'm going to create a branch of octopress-hooks that translates hooks to be based on this. If my tests pass. :+1: Let's see how long this takes :)

**Comment by envygeeks** [comment]
Can we focus more on what we need and less on what we have?

**Comment by imathis** [comment]
@envygeeks I'm focusing on what I need. I built the features in Octopress hooks because I needed them. If I can keep that functionality, I'll be good to go. Otherwise I'll have to keep monkey patching my own hooks.

**Comment by envygeeks** [comment]
@imathis Well if you run into trouble then please do post the hooks that are failing and where they originate and are used at so I might be able to help you two come to compromise by possibly reworking both sides to see if we can fit your hooks into his hook system or if we need to truly have him change his hook system.

**Comment by imathis** [comment]
@envygeeks Thanks. I don't anticipate having to rework this system. I think most of the stuff I'm doing should map to these pretty well. I sure like this code better though. I do have one question though. The way I originally wrote the hooks, they were extended from `Jekyll::Plugin`. This meant I could set priority for each hook, a feature I've found pretty useful for some of the plugins I've written. As far as I can tell, I'd set the priority on the plugin. If I needed different priority hooks, I'd just have to create separate plugins for them.

**Comment by imathis** [comment]
So far this is working beautifully. I've had one issue, though. After I set up a test for page hooks, I discovered that the incremental regeneration cache was preventing the hook from affecting the page. I cleared the cache and then it worked. I have a feeling that, "Have you tried deleting your Jekyll cache?" may become a common troubleshooting question.

**Comment by imathis** [comment]
I've rewritten my Octopress Hooks tests to use this instead of Octopress Hooks and it's awesome. The two things I lack, which I'd really like are:

1. Collection hooks.
2. All hooks (a way to write a hook that runs on any page type (Page, Post, Collection)

I definitely think Collection hooks are necessary, "All Hooks" are somewhat less necessary, but in practice the ability to tell Jekyll to process all content, avoids duplication and reduces bugs in plugins. I'm not sure how you'd implement it, but it would be really useful.

Thanks so much for this. It's way nicer to use than Octopress Hooks. :+1:

**Comment by alfredxing** [comment]
Incremental regeneration doesn't render or write pages/posts if they aren't changed. We can move hooks like `post_render` and `post_write` so they get triggered, or add new hooks that get called even when posts/pages are not rendered.

**Comment by parkr** [comment]
> "All Hooks" are somewhat less necessary, but in practice the ability to tell Jekyll to process all content, avoids duplication and reduces bugs in plugins

Eventually `Document` will be the main class; `Page` and `Post` will no longer exist at that point. Adding a hook to `Document` will take care of this. What will be harder, is getting more specific. How do you attach a filter to just one collection? Have the method only run if the doc's collection is X? Perhaps that will work for now.

**Comment by stevecrozz** [comment]
@imathis : @envygeeks and I were just talking about the priority idea. In order to do it the way that seems to make the most sense to us, we'll need a way to apply a name to a hook. So we have some changes to this PR in mind to support this. Roughly, hooks would go back to being classes, as they are in octopress-hooks, and they would implement a 'run' method and have a name (the name being the name of the class):

```ruby
class MyHook
  def run(page); end
end

class MyOtherHook
  def run(page); end
end

Jekyll::Hooks.register Jekyll::Page, :post_render, MyHook

# insert one hook before another, hook specified as a symbol here in case MyHook
# is not yet defined (due to not yet having been loaded)
Jekyll::Hooks.register Jekyll::Page, :post_render, MyOtherHook, before: :MyHook
```

I agreed to start working on this idea. As always I do appreciate feedback on this. Let me know if this looks good or bad to any of you.

@imathis I have no objections to collection hooks. If you can contribute to the cucumber tests / implementation / docs here I think we can include those.

**Comment by imathis** [comment]
I'd be happy to contribute the Collections stuff. 

On the priority, I'm looking for a way of nudging global priority like how Jekyll Plugins have the `:highest, :high, :normal, :low, :lowest`. I use this for [octopress-paginate](https://github.com/octopress/paginate/blob/master/lib/octopress-paginate/hooks.rb#L4) to ensure that other hooks have run their transformations before the paginate plugin merges the page's payload with the pagination values.

I use them again in [octopress-multilingual](https://github.com/octopress/multilingual/blob/master/lib/octopress-multilingual/hooks.rb#L4) where I try to merge page and site payloads before other hooks, so that other hooks can work with multilingual features.

I'm sure users could manually set priority by setting the load order of their gems, but it would be ideal if people didn't have to think about load order in order to use a plugin, but that would be the fallback strategy if plugins had competing priority which caused a conflict. Additionally being able to set priority like this allows user plugins written in the `_plugins` directory to participate in the priority chain.

Do you think this will take a lot of effort to do? Does it seem reasonable?

**Comment by envygeeks** [comment]
Doing it your way would drastically simplify the way the code has to work compared to the way we discussed last night where we have named hook points and a before or after, if we just have a numbered priority set then sure it's tons easier to select on.

**Comment by stevecrozz** [comment]
It would certainly be an easier change if hooks can remain as plain Ruby callables and we don't wrap them in classes. Should I just add a 'priority' keyword argument to `Jekyll::Hooks::register`?

**Comment by envygeeks** [comment]
@stevecrozz I think so, easier to do, less work and suites his needs without breaking much!

**Comment by stevecrozz** [comment]
One more thing. I'm not certain the names `:highest, :high, :normal, :low, :lowest` are clear because I can't tell if high priority means it should run earlier or later than the others. I have another suggestion which is `:earliest, :early, :normal, :late, :latest`. Any thoughts on these? Am I alone in my confusion?

**Comment by parkr** [comment]
The problem with all of this is really that you can't know the way the hooks will be run, e.g. what if two plugins have `:earliest` or `:highest`? Which one is run first?

Maybe we should order them more systematically, like with ints or something.

**Comment by envygeeks** [comment]
We already had an elaborate plan for not only out-of-order plugin priority but generic plugin priority which was what @stevecrozz mentioned yesterday, if what we want is the ability to insert yourself before another hook then the plan from yesterday is better than the plan from today.

**Comment by imathis** [comment]
@parkr: I'm fine with integer ranking or symbols. The Jekyll priorities are [currently number based](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/plugin.rb#L3), but use symbols as human friendly labels.

@envygeeks: Thinking about it more, I don't see your solution from yesterday working out very well. In the case of my multilingual plugin, I really do want it to run before most plugins, setting it to `:high` priority ensures that I can do that. This means that other plugins don't have to be aware of multilingual features to work properly. It would be unfortunate if all other plugins had to *opt-in* by registering their hooks to run after octopress-multilingual.And of course it would be impossible for me to know which plugin should be run after and keep updating my plugin, adding them to the list of hooks that should be registered after.

I prefer an agnostic global order approach.

**Comment by envygeeks** [comment]
@imathis your multilingual plugin shouldn't be a hook, it should be a generator tbh.

**Comment by envygeeks** [comment]
Or, if it's modifying things like the nav, why does it need to run first or before everything? I would assume that as translator it would want to run absolutely last so that it has access to the final modified source no?

**Comment by imathis** [comment]
@envygeeks Perhaps it could be done with a generator, I don't know. Will a generator let me merge the payload of the site and all the pages? As far as I understand, I need hooks for that. But this is an aside.

What do you think about a global priority? I know it's more of a club than a scalpel, but it also seems they could coexist. People could use the simple global priority system, or if they know of another plugin's hooks they want to proceed or follow after, they could use that instead. This way hooks could be shipped with a simple priority system, but as validating use-cases arise, your more advanced priority system could be added.

**Comment by envygeeks** [comment]
I wouldn't think of it as a global priority but in that scenario I would vote for it to just be :top or :bottom, you either insert yourself at the top of the list or you insert yourself at the bottom of the list if you don't want to hook before or after a specific plugin (with the default being :bottom) that would have already been the case to begin with (always inserting at the bottom) but inserting at the top might require an extra conditional on the sort but that's no big deal.

**Comment by imathis** [comment]
I'm fine with a three level priority system. `:top`, `:normal`, `:bottom`. I think it's useful to be able to indicate a higher priority as well as a lower priority than normal.

**Comment by pathawks** [comment]
It’d be neat if the labels could have *slightly* more descriptive names.

ie: I know my site feed needs to wait until posts have been generated.

**Comment by stevecrozz** [comment]
I'm willing to implement whatever we end up with as a consensus, but I'm starting to think the right solution is the simplest one: Load plugin hooks in the order you want them to run.

This could pose a problem only when order matters and you don't control the order. But the other solutions have their own problems too, and comparatively this problem seems easy to understand and manage.

**Comment by parkr** [comment]
> Load plugin hooks in the order you want them to run.

@stevecrozz How would you achieve this? Some are required in the source code here, some in other plugins, etc. All you have control over is your Gemfile and your config `gems` directive. I'm worried that won't work nicely.

**Comment by parkr** [comment]
@stevecrozz Let's have Hooks inherit from `Jekyll::Plugin` so they can be given a priority. Once all the plugins have been required, sort them accordingly.

**Comment by stevecrozz** [comment]
Personally, I do it like this: https://github.com/stevecrozz/lithostech.com/blob/master/_plugins/ext.rb

But my approach certainly has its limitations and I'm fine with doing it differently. @parkr, can we chat in real-time about this on IRC? I'm stevecrozz on freenode...

**Comment by parkr** [comment]
@stevecrozz I can't do today, but perhaps Sunday or Monday night? Real-time is tough. Feel free to express here – async is easier :)

**Comment by stevecrozz** [comment]
I just don't have a clear picture of what you're asking for. I have a lot of questions, but maybe the easiest way for you to explain would be for you to write the documentation for how you want this to work. Would you be willing to do that?

**Comment by parkr** [comment]
Hooks should be able to be given a priority. It would make sense to register a hook which is its own subclass of `Jekyll::Plugin`.

```ruby
class MyFunPlugin << Jekyll::Hook
  priority :highest # run me first

  def pre_render(site)
    # do mah thang
  end
end
```

Maybe?

**Comment by parkr** [comment]
@stevecrozz What do you think?

**Comment by stevecrozz** [comment]
@parkr It looks good. Did you want hooks to be subclasses of Jekyll::Hook or Jekyll::Plugin? And in your example `MyFunPlugin`, how would you specify that this is a Page hook and not a Site hook?

**Comment by stevecrozz** [comment]
Also, if we have one class per hook, it seems cleaner to make the method name something generic like 'run'. Since all plugins are not hooks, I'm assuming you want to stick with subclassing Jekyll::Hook which can be both hook declaration and registration in one step. An instance of MyFunPlugin is never needed, so the 'run' method can be a stateless, class method.
```ruby
class MyFunPlugin < Jekyll::Hook
  priority :highest # run me first
  container Jekyll::Page
  hook :pre_render

  def self.run(page, payload)
    # do mah thang
  end
end
```

But if the only thing we really want to add to this PR is priority, we can do it pretty easily without making such a big change (easier for me to implement):
```ruby
Jekyll::Hook.register Jekyll::Page, :post_render, priority: :highest do |page|
end
```

If the hook author really wants classes, that still works:
```ruby
Jekyll::Hook.register Jekyll::Page, :post_render, priority: :highest, &MyFunPlugin.method(:run)
```

**Comment by parkr** [comment]
@imathis Any preference? :point_up:

**Comment by imathis** [comment]
I'm fine with passing the priority as an option when rendering. I like that people can write simple hooks with a single line and very few options.

Looking at this again, I have an unrelated suggestion. Rather than pass a class like `Jekyll::Page` why not accept symbols like this?

```ruby
Jekyll::Hook.register :page, :post_render do |page|
end
```

The register method can determine the appropriate classes and when Jekyll eventually makes everything a collection we can just update the hooks module to use a method which filters collection by type of page.

Also, I'd like to suggest that this could be an array of page types, meaning that hooks would be run on each, for example:

```ruby
Jekyll::Hook.register [:page, :post, :collection], :post_render do |item|
end
```

Thoughts?

**Comment by StevenBlack** [comment]
I have long experience with hooks.  Some thoughts on the [proposed implementation](https://github.com/stevecrozz/jekyll/blob/jekyll-hooks/lib/jekyll/hooks.rb).

* Consider making Hooks a self-referential class.  When Hooks *are* Hooks, then we can build pipe-and-filter, chain-of-responsibility, composite, and strategy pattern implementations with surprising ease.

In the [proposed implementation](https://github.com/stevecrozz/jekyll/blob/jekyll-hooks/lib/jekyll/hooks.rb) we register Hooks with Jekyll.

In an expanded view of that, we could **also** register Hooks with Hooks.  In so dong, chained logic structures are possible.

All that's needed for this expanded view is to add a `@hook` array to the Hook class, and we iterate the `@hook` array after the business of `self` is done.

**Sidebar:** in the past I've had much cleaner results when Hooks themselves decide whether they apply to the current context.  An easy way to implement this is wrap `Hook::call()` or `Hook::execute()` in the equivalent of `if hookAppliesToThisSituation{ ... }`.

**Comment by StevenBlack** [comment]
Another note.

I like the currently proposed implementation.

I'd like to propose a thought mode.  There are two ways of looking at hooks.

* **Parser hooks** are Hooks that modify userland content at various junctures.
* **Process hooks** — which we haven't discussed much yet — are ways the Jekyll build process could be extended and enhanced at various junctures.

Both parser hooks and process hooks should have the same interface and both should be self-referenced classes.  

In other words Hooks, regardless of whether they are parser or process hooks, can delegate to other Hooks at predictable junctures common to all Hook processes.

**Comment by imathis** [comment]
@StevenBlack I like the idea of making a Hook a self referential class. Especially since that allows a user to customize internal Hook methods for each implementation, rather than having to monkey patch Jekyll::Hooks.

**Comment by StevenBlack** [comment]
@imathis exactly!

Something like this:  Ignore the third row classes because that's a higher-order discussion but basically we put a `HookAnchor` in Jekyll and we chain userland hooks on this anchor.

**Comment by stevecrozz** [comment]
I think a good next step would be to document the proposed interface from the outside so we can all give it a thumbs up, and then I will work to provide a minimal implementation. If anyone here would like to document their preferred interface, it could help provide a good reference for these discussions. Just publish a gist with new docs and link it here if you decide to help out in this way. I'll do my best to translate this thread into some new documentation also.

In general, I'm on board with these plans, but I would recommend a lot of restraint in the first implementation. From my perspective, we just need a solid foundation to build upon, something that can grow with us as we add features, but also something that doesn't aim to address use cases that we don't have. And on that note, if you do have some real use cases that you want covered, please do consider contributing some cucumber specs. Even without making them pass, they'll be very helpful for me to review.

**Comment by parkr** [comment]
We might be getting ahead of ourselves. I'm generally :+1: on what @stevecrozz said:

> I would recommend a lot of restraint in the first implementation. From my perspective, we just need a solid foundation to build upon, something that can grow with us as we add features, but also something that doesn't aim to address use cases that we don't have.

Let's move forward with more of an MVP that fits our general use cases, then build upon that.

**Comment by envygeeks** [comment]
This shit escalated fast.

**Comment by imathis** [comment]
@parkr @stevecrozz :+1: I'm happy with proceeding without hooks being self-referential classes until we decide we need them. I really would like [this](https://github.com/jekyll/jekyll/pull/3553#issuecomment-93606719) though.

**Comment by stevecrozz** [comment]
@imathis and I just discussed this and we decided to make a few changes to this PR. This is just here so anyone watching can see what we're planning to do:

1. register hooks with owners that are symbols (not direct references to classes)
2. during registration, you can specify owner as an array to register the same hook to multiple owners
3. add optional priority during registration as a symbol (:lowest, :low, :normal, :high, :highest)
4. implement hooks for collections as they are in octopress-hooks

**Comment by envygeeks** [comment]
Why does there need to be 5 instead of 3 levels of hooking?  The wording is what gets to me really not the levels because highest compared to high makes no sense in my head as far as programming is concerned.  Why not make them numbers and support legacy names until 3.1?

**Comment by stevecrozz** [comment]
@envygeeks I'm good with that. Are you, @imathis?

**Comment by imathis** [comment]
Sure, three should be fine. Five was just mimicking the Jekyll pattern for Plugins.

**Comment by StevenBlack** [comment]
> ... levels of hooking.

Careful here.  

There's a class of parsing hooks that 1) Seed content in the `pre_` hook and 2) then unseed the content in the `post_` hook.

Bearing in mind that it's hooks, plural, then an early content-seeding `pre_` hook typically needs to be a late `post_` hook for unseeding that content.  I can't think of a single instance of an early `pre_` hook that also wants to be an early `post_` hook.

If hooks *are* hooks, as I'm advocating, then the sequence of hooks is determined in userland by the way they are composed or the sequence they are either chained-together, or loaded into a flat iterable collection.

Consider that the Jekyll pattern for plugins doesn't necessarily map to hooks.

**Comment by envygeeks** [comment]
Leaving aside you want to unseed in a one time run.... unless you have absolute control over every instance of the hook system you can't guarantee anything so your logic is broken to begin with.  You cannot ensure that every hook author will not just use :highest, so there is no point in even including it because as you just pointed out, every Jim Bob programmer is going to just use :highest to "unseed" content making 5 levels moot to begin with.

**Comment by StevenBlack** [comment]
@envygeeks I recommend you [start here](http://www.amazon.com/Design-Patterns-Object-Oriented-Software-Development/dp/0201422948/ref=sr_1_1?ie=UTF8&qid=1430326570&sr=8-1&keywords=Wolfgang+pree) and then maybe we can talk about "broken logic" with some sort of sensible basis for evaluation for the design of hookable software.

**Comment by parkr** [comment]
@StevenBlack Please don't be snarky. @envygeeks is a very good developer. 

I agree that the wording could be better. `:first`, `:unspecified`, and `:last` might make even more sense if we go with 3 levels. What do you think?

**Comment by imathis** [comment]
@parkr, I'm fine with those labels. For this other stuff, I'm staying out of it. :palm_tree:

**Comment by envygeeks** [comment]
@StevenBlack I've reviewed your code, I don't think you should be handing out advice any time soon.

@parkr @imathis my only concern is the way people tend to always behave, :highest will probably be the highest common usage because everybody wants to be important so if we end up with everybody wanting to be the most important why not just reduce the levels and you can either choose to be not-important, don't-care or just plain old important.  I like the way you labled those new labels.

**Comment by StevenBlack** [comment]
@envygeeks for the record, I wrote my first-ever .rb file 11-days ago, first commit 9-days ago.

I can tell you that extensibility in general, and hook operations in particular, are easy to get-wrong.

**Comment by imathis** [comment]
If you two want to fight, meet at the flag pole after school but keep this kind of thing out of issues.

**Comment by stevecrozz** [comment]
This should address all the feedback @imathis. Just in time for JekyllConf!

File: lib/jekyll/document.rb [hunk 2]
```diff
@@ -175,11 +175,15 @@ def destination(base_directory)
     #
     # Returns nothing.
     def write(dest)
+      Jekyll::Hooks.trigger self, :post_render
```
Comment (parkr) [inline lib/jekyll/document.rb]: Why not put this at the end of render?

File: lib/jekyll/hooks.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      },
+    }
+
+    # map of all hooks and their priorities
+    @hook_priority = {}
+
+    NotAvailable = Class.new(RuntimeError)
+
+    # register hook(s) to be called later
+    def self.register(owners, event, priority: nil, &block)
```
Comment (parkr) [inline lib/jekyll/hooks.rb]: We don't use named parameters anywhere else in the code base. Advantages of this are that you can specify the block without specifying the priority?

File: lib/jekyll/hooks.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      Array(owners).each do |owner|
+        register_one(owner, event, priority: priority_value(priority), &block)
```
Comment (parkr) [inline lib/jekyll/hooks.rb]: What if `block` is nil?

File: lib/jekyll/hooks.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
+module Jekyll
+  module Hooks
+    # Helps look up hooks from the registry by owner's class
+    OWNER_MAP = {
+      Jekyll::Site => :site,
+      Jekyll::Page => :page,
+      Jekyll::Post => :post,
+      Jekyll::Document => :document,
+    }.freeze
```
Comment (parkr) [inline lib/jekyll/hooks.rb]: What if I want to make a hook for my custom plugin? Like post_sass or something?

File: lib/jekyll/site.rb [hunk 6]
```diff
@@ -158,7 +164,6 @@ def render
         end
       end
 
-      payload = site_payload
```
Comment (parkr) [inline lib/jekyll/site.rb]: This line is non-negotiable. Totally changes the speed of regen. For one site, it went from 300 seconds to 80 seconds as a result of this line. I'd like to keep it at all costs or cache the site payload on first call.

File: lib/jekyll/hooks.rb (hunk 3 above)
Comment (envygeeks) [inline lib/jekyll/hooks.rb]: I personally don't like this usage of keyword args.  Why not have it default the value to the internal API's instead of having nil seep through and then deciding on the outside, have the outside default and the internal API force a value or collapse.  That's just my opinion: the public API's set the usage and defaults.

Comment (envygeeks) [inline lib/jekyll/hooks.rb]: Let me add, I'm not saying remove the `**kwd` I'm saying, lets build an API where the defaults surface.

File: site/_docs/plugins.md [hunk 7]
```diff
@@ -470,6 +470,220 @@ module Jekyll
...
+      <th>Called</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>
+        <p><code>:site</code></p>
+      </td>
+      <td>
+        <p><code>:reset</code></p>
```
Comment (parkr) [inline site/_docs/plugins.md]: After_reset?

**Comment by parkr** [comment]
Thanks for all your work on this, @stevecrozz!! :sparkles:

Any other thoughts, @jekyll/core and @imathis?

**Comment by envygeeks** [comment]
My only complaint is that defaults aren't on the surface API, we are seeping nil through the stack into internal API's and then setting defaults.  I think that's obscure design and we should surface every default we have on the layer that is public making everything obvious.

**Comment by parkr** [comment]
I would tend to agree, Jordon! Which defaults, specifically?

File: lib/jekyll/document.rb (hunk 2 above)
Comment (stevecrozz) [inline lib/jekyll/document.rb]: That's fine with me. It's here just to match the behavior of octopress hooks:
https://github.com/octopress/hooks/blob/master/lib/octopress-hooks.rb#L256

I'm fine with moving it though. Do you still want me to do that?

File: lib/jekyll/hooks.rb (hunk 4 above)
Comment (stevecrozz) [inline lib/jekyll/hooks.rb]: Adding an exception when block does not resond to 'call'

File: lib/jekyll/hooks.rb (hunk 5 above)
Comment (stevecrozz) [inline lib/jekyll/hooks.rb]: Good idea. I think it's safe to proceed with this frozen constant for now because its just an internal API. If we need a public API for adding hook points we can change this storage mechanism easily enough.

File: lib/jekyll/site.rb (hunk 6 above)
Comment (stevecrozz) [inline lib/jekyll/site.rb]: Not sure I understand this. The same line: ```payload = site_payload``` is just 10 lines above. If it does really need to be here, do you have another suggestion on how we can allow the :site, :pre_render hook to modify the payload?

Comment (stevecrozz) [inline lib/jekyll/site.rb]: @parkr: I probably need to clarify the problem a bit more. I removed this line because it overwrites the value of the local variable 'payload' which can be altered by site pre_render hooks. There's a bunch of possible solutions to this problem but in order to suggest one, I think I need to understand this performance issue.

File: lib/jekyll/document.rb (hunk 2 above)
Comment (parkr) [inline lib/jekyll/document.rb]: Yeah, I think it'd make more sense. :)

File: lib/jekyll/site.rb (hunk 6 above)
Comment (parkr) [inline lib/jekyll/site.rb]: Hm. I was doing this to update the contents of the payload. If it still works as expected (they're all pointers I guess) then it should be fine. I'm sorry to be cross – I thought you removed the caching! Thanks, Steve.

**Comment by envygeeks** [comment]
@parkr it was originally on line 53 of hooks.rb but it looks @stevecrozz caught what I meant and correct it so :tada: I'm liking this so far!

**Comment by imathis** [comment]
Lookin' good! I'll give another go at rebuilding some plugins based on this.

File: lib/jekyll/site.rb (hunk 6 above)
Comment (stevecrozz) [inline lib/jekyll/site.rb]: No prob. I just don't want to accidentally introduce a big performance issue

**Comment by parkr** [comment]
@gjtorikian: would you mind trying out this branch on a big site you have access to? Interested in performance change, if any. :heart:

**Comment by stevecrozz** [comment]
@imathis You could potentially author a shim to make your current octopress-hooks based plugins work with these new hooks. Here's a totally untested start at something like that.
https://gist.github.com/stevecrozz/ca294104609efd8205bb

File: lib/jekyll/site.rb (hunk 6 above)
Comment (lgp36018) [inline lib/jekyll/site.rb]: hi

Stephen Crosby <notifications@github.com> wrote:
>> @@ -158,7 +164,6 @@ def render
>>          end
>>        end
>>  
>> -      payload = site_payload
>
>No prob. I just don't want to accidentally introduce a big performance
>issue
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/jekyll/jekyll/pull/3553/files#r29559695

-- 
Sent from my Android device with Gmail Plus. Please excuse my brevity.

**Comment by imathis** [comment]
@stevecrozz I'll give it a try, thanks for the guidance.

**Comment by gjtorikian** [comment]
> would you mind trying out this branch on a big site you have access to? Interested in performance change, if any.

I kept trying it out and thought it would be a quick discovery, but kept running into problem. First, most plugins aren't Jekyll 3 compatible yet; I've logged [a](https://github.com/jekyll/jekyll-sitemap/issues/74) [few](https://github.com/jekyll/jemoji/issues/17) issues for them. But even with them commented out, I discovered so of my own plugins/hacks are incompatible, too. It'll take me some time to upgrade those, in order to be able to test performance on Jekyll 3/this branch.

**Comment by parkr** [comment]
@stevecrozz Is this ready to go?

**Comment by parkr** [comment]
Merging. Feel free to add further feedback as we approach the final release!

**Comment by imathis** [comment]
I just pushed an [octopress-multilingual pull-request](https://github.com/octopress/multilingual/pull/14) for merging Jekyll hooks support. It worked really well and was nice to integrate. Thanks for your work on this @stevecrozz and everyone else who helped! :+1:

**Comment by parkr** [comment]
Hooray!!! :tada:

**Comment by stevecrozz** [comment]
Very cool. Thanks for all the help, everyone. I know there were a bunch of other ideas up there that didn't make it into this first round but it doesn't have to end here. I'll look forward to seeing and reviewing more hook-related work.
